examples/remote-settings-cli/src/downloader.rs

```diff
@@ -0,0 +1,289 @@
+    pub(crate) records: Vec<Value>,
+}
+
+pub(crate) struct CollectionDownloader {
+    pub(crate) client: reqwest::Client,
+    pub(crate) multi_progress: Arc<MultiProgress>,
+    pub(crate) output_dir: PathBuf,
+}
+
+impl CollectionDownloader {
+    pub(crate) fn new(root_path: PathBuf) -> Self {
+        let output_dir = root_path.join(DEFAULT_OUTPUT_DIR);
+        Self {
+            client: reqwest::Client::new(),
+            multi_progress: Arc::new(MultiProgress::new()),
+            output_dir,
+        }
+    }
+
+    async fn fetch_collection(
+        &self,
+        collection: String,
+        pb: ProgressBar,
+    ) -> Result<(String, CollectionData)> {
+        // First get the last_modified from monitor
+        let monitor_url = format!(
+            "{}/buckets/monitor/collections/changes/records",
+            REMOTE_SETTINGS_URL,
+        );
+
+        let monitor_response: Value = self.client.get(&monitor_url).send().await?.json().await?;
+
+        // Find the last_modified for our collection
+        let last_modified = monitor_response["data"]
+            .as_array()
+            .context("No data array in monitor response")?
+            .iter()
+            .find(|record| {
+                record["collection"].as_str() == Some(&collection)
+                    && record["bucket"].as_str() == Some("main")
+            })
+            .context("Collection not found in monitor")?["last_modified"]
+            .as_u64()
+            .context("No last_modified in record")?;
+
+        // Now fetch the actual changeset
+        let url = format!(
+            "{}/buckets/main/collections/{}/changeset?_expected={}",
+            REMOTE_SETTINGS_URL, collection, last_modified
+        );
+
+        pb.set_message(format!("Downloading {}", collection));
+
+        let response = self
+            .client
+            .get(&url)
+            .send()
+            .await
+            .context(format!("Failed to fetch {}", collection))?;
+
+        let mut changeset: Value = response
+            .json()
+            .await
+            .context(format!("Failed to parse JSON for {}", collection))?;
+
+        // Special handling for search-config-v2
+        if collection == "search-config-v2" {
+            if let Some(changes) = changeset["changes"].as_array_mut() {
+                changes.sort_by(|a, b| {
+                    let a_type = a["recordType"].as_str().unwrap_or("");
+                    let b_type = b["recordType"].as_str().unwrap_or("");
+                    let a_id = a["identifier"].as_str().unwrap_or("");
+                    let b_id = b["identifier"].as_str().unwrap_or("");
+                    (a_type, a_id).cmp(&(b_type, b_id))
+                });
+            }
+        }
+
+        // // Handle attachments if needed
+        // if collection == "search-config-icons" {
+        //     self.download_attachments("main", &collection, &changeset["changes"])
+        //         .await?;
+        // }
```

Comment:
We probably will want this enabled for mobile to begin with. I'm not sure why it is commented out here.

Something to note, is that ideally we should only package the attachments relevant to the platform / app to avoid unnecessary package bloat - the search-config-icons collection includes JEXL filters and at the moment we only have icons for desktop in there, but we'll soon be adding them for Android & iOS as well.

Reply:
Thanks for pointing it out. The PR is not done yet but I wanted an early review. Reading and understanding the bash script from Desktop gave me a headache, so I tried to cover as much as I can. I commented it out because I was not happy how the attachements are currently stored in-tree and want to find a better way. 

Can you point me to something more tangible how these attachments "work"? How do I know which attachements are for which platform? 

I am not 100% sure how we would feature gate attachments during build. Maybe @bendk has an idea?

Reply:
Sorry, I meant to include [a link](https://searchfox.org/mozilla-central/rev/30afd6631333bf83d5c62ca48c313d96a21b4bbd/services/settings/dumps/main/search-config-icons.json#16).

Specifically, the `env.appinfo.ID == \"{ec8030f7-c20a-464f-9b0e-13a3a9e97384}\"` part has the identifier for desktop.

Reply:
Feature gating seems tricky, especially when the data is split up.  Maybe something like this could work?

```
/// Get a map of collection name -> packaged data
fn get_packaged_data(collection: &str) -> HashMap<String, PackageData> {
   let mut data = HashMap::new();
   #[cfg(feature = "android-package-data")]
   {
         data.insert("foo", Package_data::from_str(include_str!("../data/foo.json)"))
         data.insert("bar", Package_data::from_str(include_str!("../data/bar.json)"));
   }
   #[cfg(feature = "ios-package-data")]
   {
       ... 
   }
   data
}
```

The combination of this feature and splitting up the data adds some complexity maintaining the list of collections, but it doesn't seem too bad to me.

Reply:
Then we need to know which collections to load for each platform. Where do we get this list from? @Standard8 @leplatrem

Reply:
> Then we need to know which collections to load for each platform. Where do we get this list from? @Standard8 @leplatrem

This is not really our decision to make in advance. 

The RS consumers will know whether their use-case will need default data or not. Sometimes an empty list works well until the first synchronization happens.
In some use-cases, it's crucial, for example, if their code need data before first paint on new profiles. Or if the UI elements rely on records and will look bad if there was no network before the first sync.
If a collection changes often (eg. Nimbus), it makes less sense, but that's the developers' decision.

What's important is that when a developer wants to package a new collection, the procedure is crystal clear with explicit points on considerations to keep in mind (binary size, etc.).

Our first concrete use-case here is the Search config AFAIU. So let's start with this one first.